File: src/TextParser.cs
using System.Text.RegularExpressions;

namespace TextParser
{
    public class Parser(Language lang)
    {
        private Language m_lang = lang;

        private static int SkipWhitespace(string text, int pos)
        {
            var maxPos = text.Length;

            for (int c = pos; c < maxPos; c++)
            {
                var ch = text[c];

                if ((ch != ' ') && (ch != '\t') && (ch != '\n') && (ch != '\r'))
                    return c;
            }

            return maxPos;
        }

        private int FindToken(string text, string tokenName, int pos, bool otherTextInside)
        {
            int ret = -1;

            var token = m_lang.tokens[tokenName];

            switch (token.type)
            {
                case Type.GroupOneChildOnly:
                case Type.Group:
                    int lowest = int.MaxValue;
                    foreach (var child in token.nestedTokens)
                    {
                        int tmp = FindToken(text, child, pos, token.otherTextInside);
                        if (tmp == pos)
                            return pos;
                        if ((tmp > pos) && (tmp < lowest))
                            lowest = tmp;
                    }
                    break;

                case Type.GroupAllChildrenInSameOrder:
                    return FindToken(text, token.nestedTokens[0], pos, token.otherTextInside);

                case Type.SimpleToken:
                case Type.StartStop:
                case Type.StartOptStop:
                case Type.DialStartAndStop:
                    int size = text.IndexOf('\n', pos);
                    if (size < pos)
                        size = text.Length - pos;
                    else
                        size = size - pos;

                    Match match = token.startRegex.Match(text, pos, size);
                    if ((match.Success)&&(match.Length > 0))
                        return match.Index;
                    break;
            }

            return ret;
        }

        private TokenResultTree ParseToken(string text, string tokenName, int pos, bool otherTextInside, string? parentEndTokenName, string? nextTokenName)
        {
            TokenResultTree ret = new();
            string? childTokenName = null;
            Match match;

            var token = m_lang.tokens[tokenName];

            //Console.WriteLine(tokenName + " at pos " + pos);

            ret.name = tokenName;

            switch (token.type)
            {
                case Type.GroupOneChildOnly:
                    foreach (var nestedTokenStr in token.nestedTokens)
                    {
                        pos = SkipWhitespace(text, pos);

                        if (FindToken(text, nestedTokenStr, pos, otherTextInside) >= pos)
                        {
                            string? localParentEndTokenName;
                            if (token.endRegex != null)
                                localParentEndTokenName = tokenName;
                            else
                                localParentEndTokenName = parentEndTokenName;

                            var child = ParseToken(text, nestedTokenStr, pos, otherTextInside, localParentEndTokenName, nextTokenName);

                            if (ret.children == null)
                                ret.children = new();
                            ret.children.Add(child);

                            ret.size = child.position + child.size - ret.position;
                            pos = child.position + child.size;
                            break;
                        }
                    }
                    break;

                case Type.Group:
                    pos = SkipWhitespace(text, pos);
                    ret.position = pos;

                    do
                    {
                        int parentTokenPos = int.MaxValue;
                        int nextTokenPos = int.MaxValue;

                        pos = SkipWhitespace(text, pos);

                        if (parentEndTokenName != null)
                        {
                            var parentEndToken = m_lang.tokens[parentEndTokenName];
                            match = parentEndToken.endRegex.Match(text, pos);
                            if ((match.Success) && (match.Length > 0))
                            {
                                if (match.Index == pos)
                                {
                                    break;
                                }
                                parentTokenPos = match.Index;
                            }
                        }

                        if (nextTokenName != null)
                        {
                            int tmpPos = FindToken(text, nextTokenName, pos, token.otherTextInside);
                            if (tmpPos == pos)
                            {
                                break;
                            }

                            if (tmpPos >= 0)
                            {
                                nextTokenPos = tmpPos;
                            }
                        }

                        int closestChildToken = int.MaxValue;
                        childTokenName = null;

                        foreach (var nestedTokenStr in token.nestedTokens)
                        {
                            int childrenPos = FindToken(text, nestedTokenStr, pos, token.otherTextInside);
                            if (childrenPos == -1)
                                continue;

                            if (childrenPos == pos)
                            {
                                childTokenName = nestedTokenStr;
                                closestChildToken = childrenPos;
                                break;
                            }

                            if ((childrenPos < closestChildToken) && (childrenPos < parentTokenPos))
                            {
                                childTokenName = nestedTokenStr;
                                closestChildToken = childrenPos;
                            }
                        }

                        if ((childTokenName == null) || (parentTokenPos < closestChildToken) || (nextTokenPos < closestChildToken))
                            break;

                        if (token.otherTextInside == true)
                            pos = closestChildToken;

                        string? localParentEndTokenName;
                        if (token.endRegex != null)
                            localParentEndTokenName = tokenName;
                        else
                            localParentEndTokenName = parentEndTokenName;

                        var child = ParseToken(text, childTokenName, pos, otherTextInside, localParentEndTokenName, nextTokenName);

                        if (ret.children == null)
                            ret.children = new();
                        ret.children.Add(child);
                        pos = child.position + child.size;
                    } while (true);
                    if (ret.children != null)
                        ret.size = ret.children.Last().position + ret.children.Last().size - ret.position;
                    break;

                case Type.GroupAllChildrenInSameOrder:
                    int c = 0;
                    foreach (var nestedTokenStr in token.nestedTokens)
                    {
                        pos = SkipWhitespace(text, pos);

                        if (c < token.nestedTokens.Length - 1)
                        {
                            nextTokenName = token.nestedTokens[c + 1];
                        }
                        else
                        {
                            nextTokenName = null;
                        }

                        string? localParentEndTokenName;
                        if (token.endRegex != null)
                            localParentEndTokenName = tokenName;
                        else
                            localParentEndTokenName = parentEndTokenName;

                        var child = ParseToken(text, nestedTokenStr, pos, otherTextInside, localParentEndTokenName, nextTokenName);

                        if (ret.children == null)
                            ret.children = new();
                        ret.children.Add(child);

                        ret.size = child.position + child.size - ret.position;
                        pos = child.position + child.size;
                        c++;
                    }
                    break;

                case Type.SimpleToken:
                    match = token.startRegex.Match(text, pos);
                    if ((match.Success == false)||(match.Length <= 0))
                        throw new Exception("Could not match " + tokenName + " token of type SimpleToken at position " + pos);
                    if ((match.Index > pos) && (token.otherTextInside == false))
                        throw new Exception("Found junk at " + pos);
                    ret.position = match.Index;
                    ret.size = match.Length;
                    break;

                case Type.StartStop:
                case Type.StartOptStop:
                    match = token.startRegex.Match(text, pos);
                    if ((match.Success == false)||(match.Length <= 0))
                        throw new Exception("Could not match start of " + tokenName + " token of type SimpleToken at position " + pos);
                    if (((match.Index - pos) > pos) && (token.otherTextInside == false))
                        throw new Exception("Found junk at " + pos);
                    ret.position = match.Index;
                    pos = match.Index + match.Length;

                    if (token.nestedTokens != null)
                    {
                        do
                        {
                            pos = SkipWhitespace(text, pos);

                            int closestTokenPos = pos;

                            var endTokenMatch = token.endRegex.Match(text, pos);
                            if ((endTokenMatch.Success == true) && (endTokenMatch.Index == pos) && (endTokenMatch.Length > 0))
                                break;

                            int closestChildToken = int.MaxValue;
                            childTokenName = null;

                            foreach (var nestedTokenStr in token.nestedTokens)
                            {
                                int childrenPos = FindToken(text, nestedTokenStr, pos, token.otherTextInside);
                                if (childrenPos == -1)
                                    continue;

                                if (childrenPos == pos)
                                {
                                    childTokenName = nestedTokenStr;
                                    closestChildToken = childrenPos;
                                    break;
                                }

                                if ((token.otherTextInside == true)&&(childrenPos < closestChildToken)&&(childrenPos < endTokenMatch.Index))
                                {
                                    childTokenName = nestedTokenStr;
                                    closestChildToken = childrenPos;
                                }
                            }

                            if (childTokenName == null)
                                break;

                            string? localParentEndTokenName;
                            if (token.endRegex != null)
                                localParentEndTokenName = tokenName;
                            else
                                localParentEndTokenName = parentEndTokenName;

                            var child = ParseToken(text, childTokenName, closestChildToken, token.otherTextInside, localParentEndTokenName, null);

                            if (ret.children == null)
                                ret.children = new();
                            ret.children.Add(child);
                            ret.size = child.position + child.size - ret.position;

                            pos = child.position + child.size;
                        } while (true);
                    }

                    pos = SkipWhitespace(text, pos);

                    match = token.endRegex.Match(text, pos);
                    if ((match.Success == false)||(match.Length <= 0))
                    {
                        if (token.type == Type.StartOptStop)
                            break;

                        throw new Exception("Could not match end of " + tokenName + " token of type SimpleToken at position " + pos);
                    }

                    if ((match.Index > pos) && (token.otherTextInside == false))
                        throw new Exception("Found junk at " + pos);

                    ret.size = match.Index + match.Length - ret.position;
                    break;

                case Type.DialStartAndStop:
                    throw new Exception("DialStartAndStop token type not implemented yet!");
            }

            Console.WriteLine("Found " + ret.name + ",at " + ret.position + ",size " + ret.size);

            return ret;
        }

        public List<TokenResultTree> ParseText(string text)
        {
            List<TokenResultTree> ret = [];

            var maxPos = text.Length;
            int pos = 0;

            do
            {
                pos = SkipWhitespace(text, pos);

                int closestTokenPos = int.MaxValue;
                string? closestTokenName = null;

                foreach (var tokenName in m_lang.startTokens)
                {
                    int foundAt = FindToken(text, tokenName, pos, m_lang.otherTextInside);
                    if ((foundAt >= pos) && (foundAt < closestTokenPos))
                    {
                        closestTokenPos = foundAt;
                        closestTokenName = tokenName;
                        if (foundAt == pos)
                            break;
                    }
                }

                if (pos == maxPos)
                    break;

                if (((closestTokenPos > pos) || (closestTokenName == null)) && (m_lang.otherTextInside == false))
                    throw new Exception("Junk found at pos " + pos + "!");

                if (closestTokenName == null)
                    break;

                var newToken = ParseToken(text, closestTokenName, closestTokenPos, m_lang.otherTextInside, null, null);

                if (newToken.size == 0)
                    throw new Exception("newToken.size!");

                ret.Add(newToken);

                pos = newToken.position + newToken.size;

            } while (pos < maxPos);

            return ret;
        }

        public void ParseLine()
        {
            // TODO: Not Implemented ParseLine()!

            //var lines = Regex.Split(text, "\r\n|\r|\n");
        }
    }
}

File: python/textparser.py
import sys
import json
import re

class TextParser:
    def __init__(self, definitionFile):
        
        with open(definitionFile, "r") as f:
            fileContent = f.read()
        
        self.definition = json.loads(fileContent)

        if self.definition.get('version') is None:
            self.definition['version'] = '0.0'

        for token in self.definition['tokens']:
            if self.definition['tokens'][token].get('otherTextInside') is None:
                self.definition['tokens'][token]['otherTextInside'] = False

            if self.definition['tokens'][token].get('deleteIfOnlyOneChild') is None:
                self.definition['tokens'][token]['deleteIfOnlyOneChild'] = False

            if self.definition['tokens'][token].get('mustHaveOneChild') is None:
                self.definition['tokens'][token]['mustHaveOneChild'] = False

            if self.definition['tokens'][token].get('multiLine') is None:
                self.definition['tokens'][token]['multiLine'] = False

            if self.definition['tokens'][token].get('searchParentEndTokenLast') is None:
                self.definition['tokens'][token]['searchParentEndTokenLast'] = False

    def __skipWhitespace(self, text, pos):
        while pos < len(text) and (text[pos].isspace() or text[pos] == '\t' or text[pos] == "\n" or text[pos] == "\r"):
            pos += 1
        return pos

    def __parseTokenGroupOneChildOnly(self, text, token_id, token, parentToken, pos):
        raise Exception("Not implemented __parseTokenGroupOneChildOnly")

    def __parseGroup(self, text, token_id, token, parentRegex, pos):

        pos = self.__skipWhitespace(text, pos)

        ret = {}
        ret['id'] = token_id
        ret['position'] = pos
        ret['length'] = 0
        ret['children'] = []

        while True:
            endTokenPos = sys.maxsize
            pos = self.__skipWhitespace(text, pos)

            closestchildTokenPos = sys.maxsize
            closestchildTokenName = None

            if ((token.get('searchParentEndTokenLast') is None)or(token.get('searchParentEndTokenLast') is not None and token['searchParentEndTokenLast'] == False)):
                if parentRegex is not None:
                    endRegex = re.search(parentRegex, text[pos:], flags=re.IGNORECASE)
                    if (endRegex is not None):
                        endTokenPos = endRegex.regs[0][0]
                        endTokenLength = endRegex.regs[0][1] - endTokenPos
                        if (endTokenPos == 0):
                            ret['length'] = pos - ret['position']
                            break
                
            for childTokenName in token['nestedTokens']:
                childTokenPos = self.__findToken(text, pos, self.definition['tokens'][childTokenName], self.definition['otherTextInside'])
                if (childTokenPos is not None):
                    if (childTokenPos < closestchildTokenPos):
                        closestchildTokenPos = childTokenPos
                        closestchildTokenName = childTokenName
                        if (closestchildTokenPos == 0):
                            break
            
            if (closestchildTokenPos > 0) and (token.get('searchParentEndTokenLast') is not None and token['searchParentEndTokenLast'] == True):
                if parentRegex is not None:
                    endRegex = re.search(parentRegex, text[pos:], flags=re.IGNORECASE)
                    if (endRegex is not None):
                        endTokenPos = endRegex.regs[0][0]
                        endTokenLength = endRegex.regs[0][1] - endTokenPos
                        if (endTokenPos == 0):
                            ret['length'] = pos - ret['position']
                            break

            if (endTokenPos < closestchildTokenPos):
                ret['length'] = pos + endTokenPos - ret['position']
                break

            if (closestchildTokenPos == sys.maxsize):
                break

            pos += closestchildTokenPos

            child = self.__parseToken(text, closestchildTokenName, self.definition['tokens'][closestchildTokenName], parentRegex, pos)
            if (child["length"] == 0):
                raise Exception("Child token " + closestchildTokenName + " has no length!")

            ret['length'] = child['position'] + child['length'] - ret['position']

            ret['children'].append(child)

            pos = child['position'] + child['length']

        return ret

    def __parseGroupAllChildrenInSameOrder(self, text, token_id, token, parentRegex, pos):
        
        pos = self.__skipWhitespace(text, pos)

        ret = {}
        ret['id'] = token_id
        ret['position'] = pos
        ret['length'] = 0
        ret['children'] = []

        for c in range(0, len(token['nestedTokens'])):
            pos = self.__skipWhitespace(text, pos)

            childTokenName = token['nestedTokens'][c]
            childTokenPos = self.__findToken(text, pos, self.definition['tokens'][childTokenName], self.definition['otherTextInside'])
            if (childTokenPos is None):
                raise Exception("Expected " + childTokenName + " at position: " + str(pos))

            pos += childTokenPos

            if c == len(token['nestedTokens']) - 1:
                currentparentRegex = parentRegex
            else:
                nextChildTokenName = token['nestedTokens'][c + 1]
                if (self.definition['tokens'][nextChildTokenName].get('startRegex') is not None):
                    currentparentRegex = self.definition['tokens'][nextChildTokenName]['startRegex']
                else:
                    currentparentRegex = None

            child = self.__parseToken(text, childTokenName, self.definition['tokens'][childTokenName], currentparentRegex, pos)
            
            ret['length'] = child['position'] + child['length'] - ret['position']

            ret['children'].append(child)
            
            pos = child['position'] + child['length']

        return ret

    def __parseSimpleToken(self, text, token_id, token, parentRegex, pos):

        pos = self.__skipWhitespace(text, pos)

        startRegex = re.match(token['startRegex'], text[pos:], flags=re.IGNORECASE)
        if (startRegex is None):
            raise Exception("Expected " + token['startRegex'] + " at position: " + str(pos))

        if (len(startRegex.regs) < 1) or (len(startRegex.regs) > 2):
            raise Exception("Expected " + token['startRegex'] + " at position: " + str(pos))

        ret = {}
        ret['id'] = token_id
        ret['position'] = pos + startRegex.regs[len(startRegex.regs) - 1][0]
        ret['length'] = startRegex.regs[len(startRegex.regs) - 1][1] - startRegex.regs[len(startRegex.regs) - 1][0]

        return ret

    def __parseStartStop(self, text, token_id, token, parentRegex, pos, endRequired):

        if (token.get('endRegex') is not None):
            parentRegex = token['endRegex']

        pos = self.__skipWhitespace(text, pos)

        startRegex = re.match(token['startRegex'], text[pos:], flags=re.IGNORECASE)
        if (startRegex is None):
            raise Exception("Expected " + token['startRegex'] + " at position: " + str(pos))

        if (len(startRegex.regs) < 1) or (len(startRegex.regs) > 2):
            raise Exception("Expected " + token['startRegex'] + " at position: " + str(pos))

        pos += startRegex.regs[len(startRegex.regs) - 1][0]

        ret = {}
        ret['id'] = token_id
        ret['position'] = pos
        ret['length'] = 0
        ret['children'] = []

        pos += startRegex.regs[len(startRegex.regs) - 1][1]

        if ((token.get('nestedTokens') == False) or (token['nestedTokens'] is None)):
            raise Exception("Token " + token['name'] + " has no nestedTokens!")

        while True:
            endTokenPos = sys.maxsize
            pos = self.__skipWhitespace(text, pos)

            if ((token.get('searchParentEndTokenLast') is None)or(token.get('searchParentEndTokenLast') is not None and token['searchParentEndTokenLast'] == False)):
                if parentRegex is not None:
                    endRegex = re.search(parentRegex, text[pos:], flags=re.IGNORECASE)
                    if (endRegex is not None):
                        endTokenPos = endRegex.regs[len(endRegex.regs) - 1][0]
                        endTokenLength = endRegex.regs[len(endRegex.regs) - 1][1] - endTokenPos
                        if (endTokenPos == 0):
                            ret['length'] = pos - ret['position'] + endTokenLength
                            break

            closestchildTokenPos = sys.maxsize
            closestchildTokenName = None
                
            for childTokenName in token['nestedTokens']:
                childTokenPos = self.__findToken(text, pos, self.definition['tokens'][childTokenName], self.definition['otherTextInside'])
                if (childTokenPos is not None):
                    if (childTokenPos < closestchildTokenPos):
                        closestchildTokenPos = childTokenPos
                        closestchildTokenName = childTokenName
                        if (closestchildTokenPos == 0):
                            break
            
            if (token.get('searchParentEndTokenLast') is not None and token['searchParentEndTokenLast'] == True):
                if parentRegex is not None:
                    endRegex = re.search(parentRegex, text[pos:], flags=re.IGNORECASE)
                    if (endRegex is not None):
                        endTokenPos = endRegex.regs[len(endRegex.regs) - 1][0]
                        endTokenLength = endRegex.regs[len(endRegex.regs) - 1][1] - endTokenPos
                        if (endTokenPos == 0):
                            ret['length'] = pos - ret['position'] + endTokenLength
                            break

            if (endTokenPos < closestchildTokenPos):
                ret['length'] = pos - ret['position'] + endTokenPos + endTokenLength
                break

            if closestchildTokenPos == sys.maxsize:
                break

            pos += closestchildTokenPos

            child = self.__parseToken(text, closestchildTokenName, self.definition['tokens'][closestchildTokenName], parentRegex, pos)
            if (child["length"] == 0):
                raise Exception("Child token " + closestchildTokenName + " has no length!")

            ret['length'] = child['position'] + child['length'] - ret['position']

            ret['children'].append(child)

            pos = child['position'] + child['length']

            pos = self.__skipWhitespace(text, pos)

            endRegex = re.search(token['endRegex'], text[pos:], flags=re.IGNORECASE)

            if (endRegex is None) and (endRequired == True):
                raise Exception("Expected endtoken for " + token_id + " at position: " + str(pos))

            if (endRegex is not None):
                endTokenPos = endRegex.regs[len(endRegex.regs) - 1][0]
                endTokenLength = endRegex.regs[len(endRegex.regs) - 1][1] - endTokenPos
                if (endTokenPos == 0):
                    ret['length'] += endTokenLength
                    break

        return ret

    def __parseDualStartStop(self, text, token_id, token, parentRegex, pos):
        raise Exception("Not implemented __parseDualStartStop")

    def __parseToken(self, text, token_id, token, parentRegex, pos):

        if (token.get('endRegex') is not None):
            parentRegex = token['endRegex']

        pos = self.__skipWhitespace(text, pos)

        match token['type']:
            case "GroupOneChildOnly":
                return self.__parseTokenGroupOneChildOnly(text, token_id, token, parentRegex, pos)
            case "Group":
                return self.__parseGroup(text, token_id, token, parentRegex, pos)
            case "GroupAllChildrenInSameOrder":
                return self.__parseGroupAllChildrenInSameOrder(text, token_id, token, parentRegex, pos)
            case "SimpleToken":
                return self.__parseSimpleToken(text, token_id, token, parentRegex, pos)
            case "StartStop":
                return self.__parseStartStop(text, token_id, token, parentRegex, pos, True)
            case "StartOptStop":
                return self.__parseStartStop(text, token_id, token, parentRegex, pos, False)
            case "DualStartStop":
                return self.__parseDualStartStop(text, token_id, token, parentRegex, pos)

        raise Exception("Unknown token type: " + token["type"])

    def __findToken(self, text, pos, token, otherTextInside):
        match token["type"]:
            case "GroupOneChildOnly" | "Group":
                closestChildPos = sys.maxsize
                for childTokenName in token['nestedTokens']:
                    childTokenPos = self.__findToken(text, pos, self.definition['tokens'][childTokenName], otherTextInside)
                    if (childTokenPos is not None): 
                        if (childTokenPos < closestChildPos):
                            closestChildPos = childTokenPos
                if (closestChildPos == sys.maxsize):
                    return None
                return closestChildPos
            case "GroupAllChildrenInSameOrder":
                return self.__findToken(text, pos, self.definition['tokens'][token['nestedTokens'][0]], otherTextInside)
            case "SimpleToken" | "StartStop" | "StartOptStop" | "DualStartStop":
                res = re.search(token["startRegex"], text[pos:], flags=re.IGNORECASE)
                if (res is None):
                    return None            
                if (res.regs[0][1] == 0):
                    return None
                if ((otherTextInside == False)and(res.pos != 0)):
                    return None
                
                return res.regs[0][0]

        raise Exception("Fatal error: Unknown token type: " + token["type"])

    def parse(self, text):
        
        tokens = []

        pos = 0

        while pos < len(text):
            pos = self.__skipWhitespace(text, pos)
            closestTokenPos = sys.maxsize
            closestTokenName = None

            for token_name in self.definition['startTokens']:
                token = self.definition['tokens'][token_name]
                offset = self.__findToken(text, pos, token, self.definition['otherTextInside'])
                if (offset is not None):
                    if offset < closestTokenPos:
                        closestTokenPos = offset
                        closestTokenName = token_name

            if (closestTokenName is None):
                break

            pos += closestTokenPos
            
            child = self.__parseToken(text, closestTokenName, self.definition['tokens'][closestTokenName], None, pos)
            if (child["length"] == 0):
                raise Exception("Child token " + closestTokenName + " has no length!")

            tokens.append(child)        
            pos = child['position'] + child['length']

        return tokens
